### PR TITLE
Blood and tears

### PR DESCRIPTION
**Pull request by guibescos**
Before the PR the program is 170k bytes. Bloat comes mostly from dependencies `bincode` and `solana_program`. The C binary is 10k. We need to shrink it to be less than 80k.

These are the tricks we used:
- Stop using `bincode`. Switch to `bytemuck` to deserialize bpf-upgradable-loader structs. (gain : 75k)
- Compiler flags seem negligible and only change +-500bytes

This wasn't enough sadly. Next we tried:
- Drop `resize_account`. This drops some statically linked functions from `solana_program`
- Drop `update_permissions` . This drops some code from `solana_program`

To avoid this pain in the future, measuring the size of the program is added to CI.

We still want to have permissions but now we don't have `update_permissions`. The trick we use is that permissioned pubkeys can be hardcoded in the binary. This means updating permissions requires a program upgrade, which seems worse that having an instruction for that.
PermissionAccount::hardcoded() provides these hardcoded pubkeys.

Things that I tried that didn't work:
- Drop  `thiserror` (negligible)
- Drop `num_derive` (negligible)
- Drop `delete_price` and `delete_product` ( a couple of k)

### CODE REVIEW

File: program/rust/src/utils.rs [hunk 1]
```diff
@@ -80,8 +77,8 @@ pub fn check_valid_signable_account_or_permissioned_funding_account(
 ) -> Result<(), ProgramError> {
     if let Some(permissions_account) = permissions_account_option {
         check_valid_permissions_account(program_id, permissions_account)?;
-        let permissions_account_data =
-            load_checked::<PermissionAccount>(permissions_account, cmd_hdr.version)?;
+
+        let permissions_account_data = PermissionAccount::hardcoded();
         check_valid_funding_account(funding_account)?;
```
Comment (guibescos) [inline program/rust/src/utils.rs:82]: Switch from reading off the account to reading from the "hardcoded" version

**Review by tompntn — COMMENTED** [review]
Generally looks ok to me, I really don't like the hardcoded pubkeys though. There may be other code/imports we can remove instead, especially around the time machine.

File: program/rust/src/processor/resize_price_account.rs [hunk 2]
```diff
@@ -35,6 +35,7 @@ use {
 // account[0] funding account       [signer writable]
 // account[1] price account         [signer writable]
 // account[2] system program        []
+#[allow(dead_code)]
 pub fn resize_price_account(
```
[the remark this answers is not in the document]
Comment (guibescos) [reply program/rust/src/processor/resize_price_account.rs:39]: Deleting the file doesn't change anything, the compiler is already ignoring it

File: program/rust/src/processor.rs [hunk 3]
```diff
@@ -72,9 +72,9 @@ pub fn process_instruction(
         UpdTest => Err(OracleError::UnrecognizedInstruction.into()),
         SetMinPub => set_min_pub(program_id, accounts, instruction_data),
         UpdPriceNoFailOnError => upd_price_no_fail_on_error(program_id, accounts, instruction_data),
-        ResizePriceAccount => resize_price_account(program_id, accounts, instruction_data),
+        ResizePriceAccount => Err(OracleError::UnrecognizedInstruction.into()),
```
[the remark this answers is not in the document]
Comment (guibescos) [reply program/rust/src/processor.rs:75]: The compiler is already ignoring this file

File: scripts/build-bpf.sh [hunk 4]
```diff
@@ -39,6 +39,7 @@ cargo clean
 cargo build-bpf
 
 sha256sum ./target/**/*.so
+./scripts/check-size.sh
```
[the remark this answers is not in the document]
Comment (guibescos) [reply scripts/build-bpf.sh:42]: I'll add these

File: program/rust/src/tests/pyth_simulator.rs [hunk 5]
```diff
@@ -330,6 +330,7 @@ impl PythSimulator {
 
     /// Resize a price account (using the resize_price_account
     /// instruction).
+    #[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (guibescos) [reply program/rust/src/tests/pyth_simulator.rs:333]: I feel like I'd rather keep this

File: scripts/check-size.sh [hunk 6]
```diff
@@ -0,0 +1,10 @@
+ORACLE_SIZE=$(wc -c ./target/deploy/pyth_oracle.so | awk '{print $1}')
+if [ $ORACLE_SIZE -lt 81760 ]
+then
+    echo "The program fits on mainnet"
```
[the remark this answers is not in the document]
Comment (guibescos) [reply scripts/check-size.sh:4]: Thank you

File: program/rust/src/accounts/permission.rs [hunk 7]
```diff
@@ -49,6 +57,23 @@ impl PermissionAccount {
             _ => false,
         }
     }
+
+    /// Creating PDAs exceeds our current program space.
+    /// Therefore the program will use the permissions specified here.
+    /// To modify the permissions the program needs to be upgraded.
+    pub fn hardcoded() -> Self {
```
[the remark this answers is not in the document]
Comment (guibescos) [reply program/rust/src/accounts/permission.rs:64]: Title of PR ^

[the remark this answers is not in the document]
Comment (guibescos) [reply program/rust/src/accounts/permission.rs:64]: 90792 is the smallest I managed to make it while preserving upg_permissions, the logic around PDA (CPIing into the system program, seems to be very expensive).

File: program/rust/src/processor/resize_price_account.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (tompntn) [reply program/rust/src/processor/resize_price_account.rs:39]: Sure, but it's better not to have dead code lying around

**Pull request closed without merging**